### PR TITLE
feat: add site theme switching mechanism

### PR DESCRIPTION
**Pull request by jzempel**
## Description

Adds the following theme switching menu to the header. Designed to select "system" by default, but persists the user's site selection to the browser's `localStorage`.

Screenshot 2024-09-24 at 11 11 54 AM

| Light | Dark |
| --- | --- |
| Screenshot 2024-11-01 at 6 15 28 PM | Screenshot 2024-11-01 at 6 14 52 PM |

## Detail

I reviewed every site page and website feature with dark mode enabled.

## Checklist

- [x] :ok_hand: website updates are Garden Designer approved (add the designer as a reviewer)
- [x] :black_nib: copy updates are approved (add the content strategist as a reviewer)
- [ ] :link: ~considered opportunities for adding cross-reference URLs (grep for keywords)~
- [ ] :wheelchair: ~analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver~
- [ ] :memo: ~tested in Chrome, Firefox, Safari, Edge~

### CODE REVIEW

File: src/components/useColorScheme.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    colorScheme: IGardenTheme['colors']['base'];
+  }>(getState(initialState));
+
+  useEffect(() => {
+    const eventListener = () => {
+      // Listen for changes to the system color scheme
+      state.isSystem && setState(getState(undefined));
+    };
+
+    mediaQuery?.addEventListener('change', eventListener);
```
Comment (geotrev) [inline src/components/useColorScheme.ts:43]: Does this generate a new event listener every time a dependency changes?

Comment (geotrev) [reply src/components/useColorScheme.ts]: Ah shoot really? React docs seemed to indicate otherwise. 😩 but i believe it.

**Comment by sieeeeeenna** [comment]
Screenshot 2024-09-06 at 11 45 57 AM (from our meeting) anchor styling on hover in dark mode needs to be adjusted

**Comment by jzempel** [comment]
> Screenshot 2024-09-06 at 11 45 57 AM (from our meeting) anchor styling on hover in dark mode needs to be adjusted

Thanks @sieeeeeenna. I have a task to cover this (and a few other related tasks) under a separate PR that _will_ merge :wink:

**Comment by jzempel** [comment]
TBD: extract all `localStorage` and `mediaMatch` logic out of the `useColorScheme` hook. Instead follow the approach documented in https://www.joshwcomeau.com/react/dark-mode/ and initialize `useColorScheme` with the results of the `--initial-color-scheme` property stored on `window.document.documentElement.style`.

**Comment by lucijanblagonic** [comment]
@jzempel Are sun/moon icons part of this work or is that happening in a separate branch? I see we also have to define the System icon if we want to use iconography for those.

File: gatsby-ssr.js [hunk 2]
```diff
@@ -25,6 +31,9 @@ export const onRenderBody = ({ setHtmlAttributes, setHeadComponents, setPostBody
     />
   ]);
 
+  /* eslint-disable-next-line react/no-danger */
+  setPreBodyComponents([<script key="color-scheme-js" dangerouslySetInnerHTML={{ __html }} />]);
```
Comment (jzempel) [inline gatsby-ssr.js:35]: This script ensures localStorage is primed with the correct `color-scheme` value and sets the HTML background color accordingly

File: src/components/useColorScheme.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * Use of this source code is governed under the Apache License, Version 2.0
+ * found at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+import { useCallback, useEffect, useState } from 'react';
+import { IGardenTheme } from '@zendeskgarden/react-theming';
+import { ColorScheme } from './useColorSchemeContext';
+
+export const useColorScheme = (initialState?: ColorScheme) => {
```
Comment (jzempel) [inline src/components/useColorScheme.ts:12]: Along with `useColorSchemeContext`, I think this hook could be a nice addition to `@zendeskgarden/react-theming` for external standard re-use.

**Comment by lucijanblagonic** [comment]
Overall looks great. Found one bug and one nitpick. :) 

**Overlay z-index issue with `Notifications` and header.**
image

**Potential improvements**
Marking these links as external, since they are the only ones in the side nav that take you outside of the website.
image

I also identified some image assets that need to be fixed (missing background, weird border) but that will be fixed in Figma directly.

**Review by ze-flo — APPROVED** [review]
Code looks great! Awaiting final approval from Design. Nice job! 💯

**Comment by jzempel** [comment]
> Marking these links as external, since they are the only ones in the side nav that take you outside of the website.

@lucijanblagonic the update also affected the footer links – but I think the consistency is for the better

**Review by lucijanblagonic — APPROVED** [review]

**Pull request merged**